src/update_handler.c

```diff
@@ -1141,6 +1142,100 @@ static gboolean img_to_fs_handler(RaucImage *image, RaucSlot *dest_slot, const g
 	return res;
 }
 
+static gboolean img_to_mbr_partition_handler(RaucImage *image, RaucSlot *dest_slot, const gchar *hook_name, GError **error)
+{
+	gboolean res = FALSE;
+	int out_fd;
+	g_autoptr(GUnixOutputStream) outstream = NULL;
+	GFile *destslotfile = NULL;
+	GError *ierror = NULL;
+	struct mbr_partition dest_partition;
+
+	/* run slot pre install hook if enabled */
+	if (hook_name && image->hooks.pre_install) {
+		res = run_slot_hook(hook_name, R_SLOT_HOOK_PRE_INSTALL, NULL,
+				dest_slot, &ierror);
+		if (!res) {
+			g_propagate_error(error, ierror);
+			goto out;
+		}
+	}
+
+	res = r_mbr_get_non_active_boot_partition(dest_slot->device, &dest_partition, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	g_message("Found inactive partition at %lu %lu", dest_partition.start, dest_partition.size);
+
+	if (dest_partition.size != image->checksum.size) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Size of image (%ld) does not fit to slot size %ld",
+				image->checksum.size, dest_partition.size);
+		res = FALSE;
+		goto out;
+	}
+
+	g_message("Opening slot device partition %s", dest_slot->device);
+
+	destslotfile = g_file_new_for_path(dest_slot->device);
+
+	out_fd = open(g_file_get_path(destslotfile), O_WRONLY);
```

Comment:
You're right - it is almost a copy. But the reason why we have to duplicate the code is that `open_slot_device()` opens the device with flag `O_EXCL`. Since we have to operate on the device and not only on a partition, this would fail.

Reply:
gstdio is not included in `update_handler.c`.

Reply:
out_fd is closed in `copy_raw_image`